### ISSUE TITLE
A class that implements an interface cannot be deserialized

### ISSUE THREAD

**Issue by HoneyRatel** [mcgrizzz/Proton#39]
Hello, I've been using Proton a lot and I've come across a serious error.

The error I found is that the class implementing the interface cannot deserialize at the sent or the receiving end in any case.

This problem has been confirmed to occur in two major cases.
Case 1: Using a class that directly implements the interface
Case 2: When using a class that includes a class that implements an interface.

  I tested this bug several times before reporting this issue,
I realized that the problem was not caused by my misconfiguration.

  Below is the code that can cause the bug in two cases.

**Case 1 Bug:**

* RabbitMQ_Test2.java

public class RabbitMQ_Test2 implements CommandExecutor {

    @MessageHandler(namespace = "rabbitmq1234", subject = "test")
    public void onRunTest(MQTest test) {

        System.out.println("Message Received");
        System.out.println("Message : " + test.getMessage());
    }

    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label, @NotNull String[] args) {

        main.getMessenger().broadcast("rabbitmq1234", "test", new Message_Test("1234"));

        sender.sendMessage("Message Sent");
        return true;
    }

    interface MQTest {

        public String getMessage();
    }
}

* Message_Test.java

public class Message_Test implements MQTest {

    private final String message;

    public Message_Test(String message) {
        this.message = message;
    }

    public String getMessage() {
        return message;
    }
}

- **Error Message: (on Message Sender)**

[16:54:55] [Server thread/WARN]: Unexpected exception while parsing console command "messagetest1"
org.bukkit.command.CommandException: Unhandled exception executing command 'messagetest1' in plugin Plugin_Message-v1.0-SNAPSHOT
	at org.bukkit.command.PluginCommand.execute(PluginCommand.java:47) ~[patched_1.17.1.jar:git-Paper-404]
	at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:159) ~[patched_1.17.1.jar:git-Paper-404]
	at org.bukkit.craftbukkit.v1_17_R1.CraftServer.dispatchCommand(CraftServer.java:869) ~[patched_1.17.1.jar:git-Paper-404]
	at org.bukkit.craftbukkit.v1_17_R1.CraftServer.dispatchServerCommand(CraftServer.java:832) ~[patched_1.17.1.jar:git-Paper-404]
	at net.minecraft.server.dedicated.DedicatedServer.handleConsoleInputs(DedicatedServer.java:529) ~[patched_1.17.1.jar:git-Paper-404]
	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:491) ~[patched_1.17.1.jar:git-Paper-404]
	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1483) ~[patched_1.17.1.jar:git-Paper-404]
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1282) ~[patched_1.17.1.jar:git-Paper-404]
	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:319) ~[patched_1.17.1.jar:git-Paper-404]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.IllegalArgumentException: Trying to send the wrong datatype for an already defined MessageContext
	at me.drepic.proton.common.ProtonManager.broadcast(ProtonManager.java:141) ~[Proton-1.3.1.jar:?]
	at spigot.RabbitMQ_Test2.RabbitMQ_Test2.onCommand(RabbitMQ_Test2.java:22) ~[Plugin_Message-1.0-SNAPSHOT.jar:?]
	at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45) ~[patched_1.17.1.jar:git-Paper-404]
	... 9 more

**Case 2 Bug:**

* RabbitMQ_Test2.java

public class RabbitMQ_Test2 implements CommandExecutor {

    @MessageHandler(namespace = "rabbitmq1234", subject = "test")
    public void onRunTest(Message_Test test) {

        System.out.println("Message Received");
        System.out.println("Message : " + test.getInner_Message().getMessage()); //Message_Test -> Message_inner (extends Message_Tail)
    }

    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label, @NotNull String[] args) {

        main.getMessenger().broadcast("rabbitmq1234", "test", new Message_Test("test", new Message_inner("test-message")));

        sender.sendMessage("Message Sent");
        return true;
    }
}

* Message_Test.java

public class Message_Test {

    private final String type;
    private final Message_Tail inner_Message;

    public Message_Test(String type, Message_Tail message) {
        this.type = type;
        inner_Message = message;
    }

    public String getType() {
        return type;
    }

    public Message_Tail getInner_Message() {
        return inner_Message;
    }
}

* Message_Tail.java

public abstract class Message_Tail {

    public abstract String getMessage();
}

* Message_inner.java

public class Message_inner extends Message_Tail {

    private final String message;

    public Message_inner(String message) {
        this.message = message;
    }

    public String getMessage() {
        return message;
    }
}

- **Error Message (on Received Server)**

[17:22:29] [pool-10-thread-4/WARN]: java.lang.RuntimeException: Failed to invoke public spigot.RabbitMQ_Test2.Message_Tail() with no args
[17:22:29] [pool-10-thread-4/WARN]: 	at com.google.gson.internal.ConstructorConstructor$3.construct(ConstructorConstructor.java:113)
[17:22:29] [pool-10-thread-4/WARN]: 	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:212)
[17:22:29] [pool-10-thread-4/WARN]: 	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
[17:22:29] [pool-10-thread-4/WARN]: 	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
[17:22:29] [pool-10-thread-4/WARN]: 	at com.google.gson.Gson.fromJson(Gson.java:932)
[17:22:29] [pool-10-thread-4/WARN]: 	at com.google.gson.Gson.fromJson(Gson.java:897)
[17:22:29] [pool-10-thread-4/WARN]: 	at com.google.gson.Gson.fromJson(Gson.java:846)
[17:22:29] [pool-10-thread-4/WARN]: 	at com.google.gson.Gson.fromJson(Gson.java:817)
[17:22:29] [pool-10-thread-4/WARN]: 	at Proton-1.3.1.jar//me.drepic.proton.common.ProtonManager.notifyHandlers(ProtonManager.java:253)
[17:22:29] [pool-10-thread-4/WARN]: 	at Proton-1.3.1.jar//me.drepic.proton.common.RabbitMQManager.deliverCallback(RabbitMQManager.java:89)
[17:22:29] [pool-10-thread-4/WARN]: 	at Proton-1.3.1.jar//com.rabbitmq.client.impl.recovery.AutorecoveringChannel$2.handleDelivery(AutorecoveringChannel.java:586)
[17:22:29] [pool-10-thread-4/WARN]: 	at Proton-1.3.1.jar//com.rabbitmq.client.impl.ConsumerDispatcher$5.run(ConsumerDispatcher.java:149)
[17:22:29] [pool-10-thread-4/WARN]: 	at Proton-1.3.1.jar//com.rabbitmq.client.impl.ConsumerWorkService$WorkPoolRunnable.run(ConsumerWorkService.java:104)
[17:22:29] [pool-10-thread-4/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[17:22:29] [pool-10-thread-4/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[17:22:29] [pool-10-thread-4/WARN]: 	at java.base/java.lang.Thread.run(Thread.java:833)
[17:22:29] [pool-10-thread-4/WARN]: Caused by: java.lang.InstantiationException
[17:22:29] [pool-10-thread-4/WARN]: 	at java.base/jdk.internal.reflect.InstantiationExceptionConstructorAccessorImpl.newInstance(InstantiationExceptionConstructorAccessorImpl.java:48)
[17:22:29] [pool-10-thread-4/WARN]: 	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
[17:22:29] [pool-10-thread-4/WARN]: 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
[17:22:29] [pool-10-thread-4/WARN]: 	at com.google.gson.internal.ConstructorConstructor$3.construct(ConstructorConstructor.java:110)
[17:22:29] [pool-10-thread-4/WARN]: 	... 15 more

Also, this problem occurs not only for classes that implement interfaces, but also for classes that implement abstract classes.

Server Version: Paper 1.17.1
Java Version : OpenJDK 17
Plugin Version: Proton v1.3.1 (using RabbitMQ)

As a huge fan of this plugin, I hope this issue gets resolved as well as possible.

Thanks for reading.

**Comment by msm595** [comment]
Proton does not support serialization and deserialization of arbitrary subclasses - if you tell it that it will receive or send a message of type X, it will try to deserialize the message directly into an Object of type X. Because the Proton message format does not include the underlying class name, it cannot know that when you say an object is of type `MQTest` that you actually mean the message is of type `Message_Test` or some other subclass. Trying to automatically handle this case could also break some functionality of Proton (e.g. what happens if you register a handler for both `MQTest` and `Message_Test`, which handler should receive the message, or should both? Assuming they are both normal classes. Or what about if a message is received with subclass2 but only a handler is registered for subclass1. There are a lot of unknowns here). So when you tell it to create an instance of an interface, it will try to create one (rather than creating an instance of a class which implements that interface). This is the cause of the two errors you have observed.

You can implement the subclass dispatching yourself. In this example, `ComplicatedData` is completely sendable and receivable on proton, and `getInterfaceObject()` returns either an instance of SubClass1 or SubClass2.

```
    interface MyInterface {
        String getString();
    }

    static class SubClass1 implements MyInterface {
        String string1;

        @Override
        public String getString() {
            return string1;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            SubClass1 subClass1 = (SubClass1) o;
            return Objects.equals(string1, subClass1.string1);
        }

        @Override
        public int hashCode() {
            return Objects.hash(string1);
        }
    }

    static class SubClass2 implements MyInterface {
        String string2;

        @Override
        public String getString() {
            return string2;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            SubClass2 subClass2 = (SubClass2) o;
            return Objects.equals(string2, subClass2.string2);
        }

        @Override
        public int hashCode() {
            return Objects.hash(string2);
        }
    }

    static class ComplicatedData {

        public int a;
        public float b;
        public String c;
        public List<Character> d;

        public Object interfaceObject;
        public String interfaceType;

        public ComplicatedData(int a, float b, String c, List<Character> d, MyInterface myInterface) {
            this.a = a;
            this.b = b;
            this.c = c;
            this.d = d;
            this.interfaceObject = myInterface;
            this.interfaceType = myInterface.getClass().getName();
        }

        public MyInterface getInterfaceObject() throws ClassNotFoundException {
            Gson gson = new Gson();
            Class clazz = Class.forName(this.interfaceType);
            String objectString = gson.toJson(this.interfaceObject);
            return (MyInterface) gson.fromJson(objectString, clazz);
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            ComplicatedData that = (ComplicatedData) o;
            try {
                return a == that.a && Float.compare(that.b, b) == 0 && c.equals(that.c) && d.equals(that.d) && this.getInterfaceObject().equals(that.getInterfaceObject());
            } catch (ClassNotFoundException e) {
                throw new RuntimeException(e);
            }
        }

        @Override
        public int hashCode() {
            try {
                return Objects.hash(a, b, c, d, this.getInterfaceObject());
            } catch (ClassNotFoundException e) {
                throw new RuntimeException(e);
            }
        }
    }
```

**Comment by HoneyRatel** [comment]
First of all, thank you for your accurate and kind reply.

After reporting this issue, I retested the code in various ways and realized that the problem was caused by a slightly different serialization/deserialization method in the proton library than I had originally thought.

Therefore, in order to solve this problem, the correct way is to send a arbitrary class to object as in the example written By sir msm595 and then deserialize it in the specified way at the receiving end of the message.

This problem will be a useful solution for anyone who encounters the same problem while trying to serialize/deserialize in this plugin as well as other messaging plugins in the future.

Once again, thank you so much for your quick response in kind.
